server/src/resources/base/mod.rs

```diff
@@ -1,40 +1,151 @@
+use crate::postgres::{PsqlFieldType, PsqlJsonValue};
 
-/// Validation Error
-#[derive(Debug)]
-pub struct ValidationError {
-    pub field: String,
-    pub error: String,
+#[derive(Clone, Debug)]
+pub struct ResourceDefinition {
+    pub psql_table: String,
+    pub psql_id_col: String,
+    pub fields: HashMap<String, FieldDefinition>,
+}
+pub trait Resource {
+    /// Allows us to implement the resource definition used for simple insert and update queries
+    fn get_definition() -> ResourceDefinition;
+    /// This function should be implemented for the resources where applicable (example implementation can be found in the flight_plan module).
+    fn get_enum_string_val(field: &str, value: i32) -> Option<String>;
 }
 
-/// Convert a prost_types::Timestamp (used by grpc)
-/// into a chrono::DateTime::<Utc> (used by postgres)
-/// TODO: Can we use Traits for or simplify in another way?
-pub fn ts_to_dt(ts: &Timestamp) -> Result<DateTime<Utc>, ArrErr> {
-    let seconds = ts.seconds;
-    let nanos: u32 = match ts.nanos.try_into() {
-        Ok(n) => n,
-        Err(e) => return Err(ArrErr::from(e)),
-    };
+#[derive(Clone, Debug)]
+pub struct FieldDefinition {
+    pub field_type: PsqlFieldType,
+    mandatory: bool,
+    internal: bool,
+    default: Option<String>,
+}
+impl FieldDefinition {
+    pub fn new(field_type: PsqlFieldType, mandatory: bool) -> Self {
+        Self {
+            field_type,
+            mandatory,
+            internal: false,
+            default: None,
+        }
+    }
+    pub fn new_internal(field_type: PsqlFieldType, mandatory: bool) -> Self {
+        Self {
+            field_type,
+            mandatory,
+            internal: true,
+            default: None,
+        }
+    }
 
-    let dt = NaiveDateTime::from_timestamp_opt(seconds, nanos);
-    match dt {
-        Some(dt) => Ok(DateTime::<Utc>::from_utc(dt, Utc)),
-        None => Err(ArrErr::Error(
-            "failed to convert timestamp to date time".to_string(),
-        )),
+    pub fn is_mandatory(&self) -> bool {
+        self.mandatory
+    }
+    pub fn is_internal(&self) -> bool {
+        self.internal
+    }
+    pub fn has_default(&self) -> bool {
+        self.default.is_some()
+    }
+    pub fn set_default(&mut self, default: String) -> Self {
+        self.default = Some(default);
+        self.clone()
+    }
+    pub fn get_default(&self) -> String {
+        if self.has_default() {
+            self.default.clone().unwrap()
+        } else {
+            panic!("get_default called on a field without a default value");
+        }
+    }
+}
+
+impl TryFrom<Id> for Uuid {
+    type Error = ArrErr;
+    fn try_from(id: Id) -> Result<Self, ArrErr> {
+        Uuid::try_parse(&id.id).map_err(ArrErr::UuidError)
+    }
+}
+
+impl From<PsqlJsonValue> for Vec<i64> {
+    fn from(json_value: PsqlJsonValue) -> Vec<i64> {
+        let cargo_weight_g = json_value.value.as_array().unwrap();
```

Comment:
interesting variable names here

Reply:
Ha, seems like I didn't change them after moving this function to this implementation